### DISCUSSION TITLE
No logs in zigbee2mqtt/bridge/logging and web interface on dev branch

### DISCUSSION THREAD

**Discussion by schauveau** [Koenkk/zigbee2mqtt#22345] [General]
I recently noticed that when using my dev build of Z2M, the logs are not anymore published in  `zigbee2mqtt/bridge/logging`
 
They do not appear either in the web interface.

I know that some changes were made recently in the log api (I recently got some debug related build errors and also `log.txt` was changed to `log.log`) but I am wondering if I am missing something. A new option to enable logs in MQTT maybe?

**Comment by Koenkk** [comment]
The logs are still published, except the `debug` ones (for performance reasons)

**Reply by schauveau** [reply]
Ok! And I see that the `MQTT publish` messages are now in `debug` instead of `info`.
Also, most  of the `info` logs produced during startup and shutdown are not published in `zigbee2mqtt/bridge/logging. 

During normal operations, the only `info` that are published in `zigbee2mqtt/bridge/logging`  are 
 
```
Zigbee2MQTT:info 2024-04-29 23:48:36: Started frontend on port 8080
Zigbee2MQTT:info 2024-04-29 23:48:36: Zigbee2MQTT started!
```

**Comment by schauveau** [comment]
Also, looking at the huge amount of debug logs that are produced, I wonder how inefficient that is even when debug logs are not enabled.

```
[2024-04-29 23:57:48] debug: 	zh:zstack:znp: AREQ: <-- AF - incomingMsg - {"groupid":0,"clusterid":61184,"srcaddr":5524,"srcendpoint":1,"dstendpoint":1,"wasbroadcast":0,"linkquality":60,"securityuse":0,"timestamp":15011668,"transseqnumber":0,"len":13,"data":{"type":"Buffer","data":[9,221,2,0,220,112,2,0,4,0,0,9,49]}}
[2024-04-29 23:57:48] debug: 	zh:controller: Received payload: clusterID=61184, address=5524, groupID=0, endpoint=1, destinationEndpoint=1, wasBroadcast=false, linkQuality=60, frame={"header":{"frameControl":{"frameType":1,"manufacturerSpecific":false,"direction":1,"disableDefaultResponse":false,"reservedBits":0},"manufacturerCode":null,"transactionSequenceNumber":221,"commandIdentifier":2},"payload":{"seq":56320,"dpValues":[{"dp":112,"datatype":2,"data":{"type":"Buffer","data":[0,0,9,49]}}]},"command":{"ID":2,"parameters":[{"name":"seq","type":33},{"name":"dpValues","type":1011}],"name":"dataReport"}}
```

For example, let's consider the first line in that sequence.  It is probably emitted here  
https://github.com/Koenkk/zigbee-herdsman/blob/e90595a0b57d2f8e185744c7e64bf3f824d4f7c7/src/adapter/z-stack/znp/znp.ts#L80
but the `message` string has to be prepared here
https://github.com/Koenkk/zigbee-herdsman/blob/e90595a0b57d2f8e185744c7e64bf3f824d4f7c7/src/adapter/z-stack/znp/znp.ts#L94

So `message` is always constructed. Calling `JSON.stringify` and formatting the string are probably not very expensive operations but they are not cheap either. 

The issue here is not even that `message` is prepared before calling `logger.debug`. The problem persists even when `JSON.stringify` appears within the argument of `logger.debug`as in 
https://github.com/Koenkk/zigbee-herdsman/blob/e90595a0b57d2f8e185744c7e64bf3f824d4f7c7/src/adapter/ember/adapter/emberAdapter.ts#L3882

I know that interpreted language can apply a lot of "magical optimization tricks" but is is very likely that the arguments have to be evaluated before calling `logger.debug` and so before testing if the debug logs are enabled or not.

**Reply by Koenkk** [reply]
This was also the case before when we were still using `DEBUG=zigbee-herdsman:*` env variable for debugging. The biggest performance impact was the publishing to MQTT.

**Comment by BandBxx** [comment]
To be honest, after the update it is not clear how device “spammers” are tracked. Previously, in the Z2M console it was convenient to filter by name in the "info" mode. Now it seems that only two modes work in the add-on logs: 1. debug is too detailed, impossible to use. 2. info, which in fact do not work. Possibly a bug. But even if it works, it is not clear how to filter messages in the add-on logs.

**Comment by schauveau** [comment]
I kind of remember using a environment variable to enable debug in some zigbee-herman files.

Now, it seems that you have to do the opposite: Debug output can be filtered out using an option.  
 
https://www.zigbee2mqtt.io/guide/usage/debug.html

So add the following to get rid of the gazillion of zigbee-herman debug messages:

```yaml
advanced:
   log_debug_namespace_ignore: '^zh:'
```

I may have missed something but, unfortunately,  the environment variable does not appear to exist anymore.

**Reply by BandBxx** [reply]
Thx. But why not make a UI filter?

**Comment by schauveau** [comment]
I forgot to mention that `log_debug_namespace_ignore` seems to be very recent so you may have to update zigbee2mqtt.

**Comment by schauveau** [comment]
I just noticed something a bit counter-intuitive. The trailing `:` is not part of the message namespace. For example, the regex `^z2m:` does not disable messages such as `debug:    z2m: Received Zigbee message from ...`. 
The proper regex is `^z2m` or `^z2m$`

**Reply by BandBxx** [reply]
The average user does not need the overloaded "debug" mode at all. Return a simple working "info" mode)
